aws/resource_aws_dx_public_virtual_interface_test.go

```diff
@@ -0,0 +1,133 @@
+package aws
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/directconnect"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAwsDxPublicVirtualInterface_basic(t *testing.T) {
+	key := "DX_CONNECTION_ID"
+	connectionId := os.Getenv(key)
+	if connectionId == "" {
+		t.Skipf("Environment variable %s is not set", key)
+	}
```

Comment:
Can't we just automatically create the DX connection as part of the test here?

Reply:
The problem is that the connection needs to be in the `Active` state for any of the Virtual Interface functionality to succeed and activation of a connection requires sending a [Letter of Activation](https://docs.aws.amazon.com/directconnect/latest/UserGuide/create-connection.html#create-connection-loa-cfa) to the network/CoLo provider and manual work on their side; i.e. plenty of time and money.
I have been testing on connections that already have active but aren't used for live traffic.

Reply:
Hi @ewbankkit, we are just about to implement two new Direct Connect connections in eu-west-2, and in advance I created a test connection in a sandbox account so that I could map out the Virtual interfaces I needed. I have no intention of ever sending the LOA for this connection, but I am able to create and attach Virtual Interfaces to it, even though it is in the down state.

Reply:
@X-Guardian What state does the Virtual Interface end up in? The resource won't be successfully created unless it ends up in _available_ or _verifying_ states.

Reply:
@ewbankkit, I can create a Virtual Interface resource connected to either a DC Gateway or a VPG, and it creates successfully. Yes, it has a status of down of course, but does that matter?

Reply:
@X-Guardian When I try and create a Direct Connect Connection and then a Public Virtual Interface on that connection in the acceptance test

```hcl
resource "aws_dx_connection" "foo" {
  name      = "%s"
  bandwidth = "1Gbps"
  location  = "EqSe2"
}

resource "aws_dx_public_virtual_interface" "foo" {
  connection_id    = "${aws_dx_connection.foo.id}"

  name           = "%s"
  vlan           = 4094
  address_family = "ipv4"
  bgp_asn        = 65352

  customer_address = "175.45.176.1/30"
  amazon_address   = "175.45.176.2/30"
  route_filter_prefixes = [
    "210.52.109.0/24",
	"175.45.176.0/22"
  ]
}
```

then the test fails

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxPublicVirtualInterface_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAwsDxPublicVirtualInterface_ -timeout 120m
=== RUN   TestAccAwsDxPublicVirtualInterface_basic
--- FAIL: TestAccAwsDxPublicVirtualInterface_basic (20.73s)
	testing.go:518: Step 0 error: Error applying: 1 error(s) occurred:
		
		* aws_dx_public_virtual_interface.foo: 1 error(s) occurred:
		
		* aws_dx_public_virtual_interface.foo: Error creating Direct Connect public virtual interface: DirectConnectClientException: The specified Connection ID is not available
			status code: 400, request id: 0fb430bb-6c2f-11e8-8b43-895a753e5994
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	20.744s
```

The connection is in the _ordering_ state when the VIF is created on it.

Reply:
@ewbankkit, when I have created a Direct Connect connection, it can take anything from 15 minutes to several hours to move on from the 'ordering' state, depending upon the partner chosen. Can you split your dx connection test from your virtual interface test?

Reply:
@X-Guardian Yes, that's why the acceptance tests require the environment variable `DX_CONNECTION_ID` to be set to the ID of the connection created elsewhere.